### PR TITLE
First draft of py-allotax

### PR DESCRIPTION
**Pull request by amafehr**
# Summary

This first pass implements everyone's templated work to have python's subprocess run allotax charting, intercept its svg, and save it to file. 

# Scope of PR

The purpose of this pull request is just to agree on organization (i.e., this is the structure of how we want to use methods in py-allotax and produce output).

## Description
- Moves Carter's doughnut chart proof-of-concept to a new directory.
- Adds directories where users can put data, files to convert to the correct format, and have output data be saved.
- Adds code to `src/` to run allotax charting, intercept the `svg`, and save it to file!
- Adds a software diagram that will get updated as things change. 
- Documents current install instructions and tool design decisions in `READ.md` and `DECISIONS.md`, respectively.

## Test
Test by running:
```
cd src &&
python3 generate_svg.py data/boys_2022.json data/boys_2023.json output_charts/test.pdf "0.17"
```
You should see saved an `.html` file and a `.pdf` file, both showing the full set of allotaxonometer plots when opened. Please comment on any issues in output.

## Notes and known issues
- We don't have a `requirements.txt` or setup yet. This will come along in the next one. Please reach out if this make install difficult.
- an install of allotaxonometer=1.1.12 is required along with changes to package imports and adding `svg` as args to all the plotting modules. For now, I implemented @jstonge's idea to branch a branch in `allotaxonometer` for changes needed here, and am including instructions to `npm install jstonge/allotaxonometer#py-allotax-experimental`
    - This current set up lets us pull changes from the `main` library into our tool's version based on need.
- need node dependencies (@carterwward if we go the binaries route then we need the binaries and a script to auto install the right ones for your OS. See `DECISIONS.md`).
    - [Discussion of binaries option](https://stackoverflow.com/questions/49953785/which-options-do-exist-for-defining-a-python-package-with-node-js-dependencies)
- need setup and general packaging accoutrements (this will be a separate, future PR).
- need to solidify the data format this takes. It seems like json--I'd been a little lost in details and did some unnecessary things I need to clean up.

### CODE REVIEW

**Comment by amafehr** [comment]
merged in changes from my other branch and am working on cleaning up so people can review.

**Comment by amafehr** [comment]
All done my envisioned clean-up for this PR. One clarification to keep in mind is right now we're serializing results to HTML and generating a PDF. I did quite a bit of exploration on methods to turn HTML into any image (see first the issue I opened then if desired `DECISIONS.md` and `pandoc_weasyprint_attempts`).

**Comment by carterwward** [comment]
to manage python dependencies we could use https://pdm-project.org/latest/

File: proof_of_concept/visualization.svg
Comment (carterwward) [inline proof_of_concept/visualization.svg:1]: can be deleted

File: src/index.html
Comment (carterwward) [inline src/index.html:1]: I think we want to remove any lingering html files, but maybe I am jumping the gun here

**Review by jstonge — COMMENTED** [review]
trying to follow naively the README and I get
```
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
when running 
```zsh
python generate_svg.py ../data/boys_2022.js ../data/boys_2023.js output.svg "0.33"
```

File: DECISIONS.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+## Decision
+Regarding the ideation below, we ended up going with ...
+
+## Design
+
+### Goals
+1. User accessibility for the use case of integrating the graphing method in computational research.
+    - This goal means ease of ease of use and ease of installation into their environment.
+    - **We agree this is the main goal of the project.**
```
Comment (jstonge) [inline DECISIONS.md:11]: is this a question? If so, yes we agree.

File: DECISIONS.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+1. Maintainability
+    - It is easy to make updates to the code, hopefully from its main `npm` package.
+    - One realism we already foresee is that the `npm` package exists to be extensible for use cases beyond the current one--the current use case requires some tailoring to intercept the graph `svg` and save it to a file, potentially meaning we store a version of the package in this repo to make adjustments for serverless graph creation.
+1. Performance
+    - The graphing method runs headlessly and should be able to handle large data sets as well as be integrated into larger analyses, pipelines, parallel processing, etc.
```
Comment (jstonge) [inline DECISIONS.md:16]: I guess one question is how large is large? Maybe we could run some benchmarks on twitter data.

File: src/index.html
Comment (jstonge) [reply src/index.html:1]: I think you are jumping the guns! If im right, we are loading that html file in particular, then target it with d3js to have the structure that we want (to assemble the multiple plots into the allotaxonograph). See lines `line 13` in `generate_svg_minimum.js`.

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,9 @@
 venv
 node_modules
 package-lock.json
+.DS_Store
+*output_charts/
+*data/
```
[the remark this answers is not in the document]
Comment (amafehr) [reply .gitignore:6]: I purposely added the data/ and output_charts/ so we have tests that can be run. I thought to .gitignore them for future stuff users addd.

File: src/index.html
Comment (amafehr) [reply src/index.html:1]: Yes, the `index.html` files are the template put into `JSDOM`

**Comment by amafehr** [comment]
> trying to follow naively the README and I get
> 
> ```
> json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> ```
> 
> when running
> 
> ```shell
> python generate_svg.py ../data/boys_2022.js ../data/boys_2023.js output.svg "0.33"
> ```

@jstonge Can you try the testing example listed in the description?

**Comment by amafehr** [comment]
> to manage python dependencies we could use https://pdm-project.org/latest/. Would be happy to get that working if folks interested

@carterwward I am happy for you to take on all the requirements and packaging stuff if you desire. I know I owe an actual list of JS and python package requirements to do so; I was just waiting to see on a few directions.

**Comment by jstonge** [comment]
> > trying to follow naively the README and I get
> > ```
> > json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > when running
> > ```shell
> > python generate_svg.py ../data/boys_2022.js ../data/boys_2023.js output.svg "0.33"
> > ```
> 
> @jstonge Can you try the testing example listed in the description?

Using the following it works!
`python generate_svg.py convert/boys_2022.json convert/boys_2023.json output_charts/test.pdf "0.17"`

I guess we need to have chrome installed to work right? While running the function, it asked me permission for chrome to access my folder.

**Comment by jstonge** [comment]
I think it might be a good place to do the PR and mark that as success! We have something to show (although i might try to do the contour lines as soon as that PR is merged, but this will be in the allotaxonometer experimental branch).

**Comment by amafehr** [comment]
> > > trying to follow naively the README and I get
> > > ```
> > > json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > when running
> > > ```shell
> > > python generate_svg.py ../data/boys_2022.js ../data/boys_2023.js output.svg "0.33"
> > > ```
> > 
> > 
> > @jstonge Can you try the testing example listed in the description?
> 
> Using the following it works! `python generate_svg.py convert/boys_2022.json convert/boys_2023.json output_charts/test.pdf "0.17"`
> 
> I guess we need to have chrome installed to work right? While running the function, it asked me permission for chrome to access my folder.

Huzzah! Yes unfortunately the current HTML>PDF tool relies on having a chrome driver--so it will detect if you already have Chrome and use that. This method is ideally what we want to replace (the thing I've been flipflopping about).

**Comment by amafehr** [comment]
@jstonge if you're happy, mark as "approved" and I'll wait on Carter (and until meeting later).

**Review by jstonge — APPROVED** [review]
Here's a summary of my understanding of this PR.

The goal of `py-allotax` is to make available to the broader community the allotaxonograph. To do that, we rely [allotaxonometer.js](https://github.com/jstonge/allotaxonometer/tree/py-allotax-experimental) to do the calculation and the charting. We have tried to strip down`py-allotax` as much as we could. At the moment, we found the following solution
1. Install allotaxonometer#py-allotax-experimental branch via npm.
2. Using `node.js`, run `subprocess.run(command, capture_output=True, text=True)` (`src/generate_svg.py`) to run the javascript code. We provide input (data) to that script and commands for parameter values. 
3. the above command runs `generate_svg_minimum.js`, a javascript file that create a special dom via `JSDOM()`. In that `JSDOM()`, we use a `.html` template to organize the different charts, which we target using `d3js`.
4. Once we have charter the graph, we serialize the DOM into SVG [?]. We then convert the SVG to HTML and HTML to PDFs via different libraries.

This is a nice proof of concept for anybody that want to convert JS code into programmatic Python. We can streamline the process more (maybe getting rid of `node.js` and dependencies on a driver, but this is a good start.

File: DECISIONS.md (hunk 2 above)
Comment (carterwward) [reply DECISIONS.md:16]: If we are rendering the html in some hidden browser I have a feeling that will be the bottle neck

File: package.json
Comment (carterwward) [inline package.json:1]: Is we have our own build script, this should be enough to manage the js install. Then we can just use pdm for the python dependencies within the same script.

**Comment by carterwward** [comment]
> I think it might be a good place to do the PR and mark that as success! We have something to show (although i might try to do the contour lines as soon as that PR is merged, but this will be in the allotaxonometer experimental branch).
> 
> edit: maybe having the dependencies fix would be nice, actually @carterwward .

I think that can be a separate pr. We will still have to do packaging and stuff before a consumable release so I think it makes sense to take the win and lump that in with build stuff.

File: src/generate_svg.py
Comment (carterwward) [inline src/generate_svg.py:1]: What is this file being used for now? I see the python module calls minimum.

File: DECISIONS.md (hunk 2 above)
Comment (amafehr) [reply DECISIONS.md:16]: I encourage creating an issue if there's definitely a 'big data' use case we want to benchmark for.

File: src/generate_svg.py
Comment (amafehr) [reply src/generate_svg.py:1]: I think this is resolved, but reiterating: `generate_svg.py` is the python end. `generate_svg_minimum.js` is a dumb name to indicate this is a minimum working script to have the JS do what we want (run JS scripts to construct, use the DOM, serialize, return HTML).

**Pull request merged**